kvbc/src/v4blockchain/detail/keys_history.cpp

```diff
@@ -0,0 +1,226 @@
+// Concord
+//
+// Copyright (c) 2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "util/endianness.hpp"
+#include "v4blockchain/detail/keys_history.h"
+#include "v4blockchain/detail/column_families.h"
+#include "v4blockchain/detail/blockchain.h"
+#include "rocksdb/details.h"
+#include "build/kvbc/cmf/categorized_kvbc_msgs.cmf.hpp"
+
+using namespace concord::kvbc;
+namespace concord::kvbc::v4blockchain::detail {
+
+// helper type for the visitor
+template <class... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+// explicit deduction guide
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+
+BlockId KeysHistory::last_block_{0};
+
+KeysHistory::KeysHistory(
+    const std::shared_ptr<concord::storage::rocksdb::NativeClient>& native_client,
+    const std::optional<std::map<std::string, concord::kvbc::categorization::CATEGORY_TYPE>>& categories)
+    : native_client_{native_client}, category_mapping_(native_client, categories) {
+  if (native_client_->createColumnFamilyIfNotExisting(v4blockchain::detail::KEYS_HISTORY_CF,
+                                                      KHCompactionFilter::getFilter())) {
+    LOG_INFO(V4_BLOCK_LOG, "Created [" << v4blockchain::detail::KEYS_HISTORY_CF << "] column family for keys history");
+  }
+}
+
+void KeysHistory::addBlockKeys(const concord::kvbc::categorization::Updates& updates,
+                               const BlockId block_id,
+                               storage::rocksdb::NativeWriteBatch& write_batch) {
+  LOG_DEBUG(V4_BLOCK_LOG, "Adding keys of block [" << block_id << "] to the keys history CF");
+
+  last_block_ = block_id;
+  for (const auto& [category_id, category_updates] : updates.categoryUpdates().kv) {
+    std::visit(overloaded{[category_id = category_id, &write_batch, &block_id, this](const auto& updates) {
+                            std::vector<std::string> updates_keys(updates.kv.size());
+                            for (const auto& [k, _] : updates.kv) {
+                              (void)_;  // unused
+                              updates_keys.push_back(k);
```

Comment:
The code is good, and you did a nice trick here, but I think it's better to do as done in the latest_keys handling i.e. overload each category to accept its update.
First, there is redundant copying of data e.g. push back on the vector, and in the handleUpdateImp as well, you can use getSliceArray from the latest_keys to compose your key minimal copies.
and also it will enable us to filter by configuration categories that we don't want to store

Reply:
Ok, thanks